### PR TITLE
Initial support for Blueprints

### LINKED ISSUE #12: Add Blueprint support

### PR DESCRIPTION
**Pull request by boydgreenfield**
All of the tests should pass, and I've tested moving an app from using Potion w/o Blueprints to Blueprints with no apparent adverse effects.

**Note**: One change here is that `Api.__init__` now joins unnamed routes with `_` rather than `.` (see https://github.com/refgenomics/potion/blob/blueprints/flask_potion/__init__.py#L159-L162) if the `Api` is initialized with a Blueprint. This should potentially be simplified to always use `_` (`.` does not work with Blueprints).

**Caveats**: I have _not_ experimented with registering the same Blueprint multiple times (and if that has any unintended consequences) or registering multiple Blueprints (which this comment indicates may require a few changes https://github.com/biosustain/potion/blob/master/flask_potion/__init__.py#L83).

Closes #12.

@lyschoening – would appreciate your review and any further suggested changes.

### CODE REVIEW

**Comment by boydgreenfield** [comment]
I'm signing off for the evening here, but can also add a commit updating the documentation or a separate PR as appropriate tomorrow or beginning of next week (holiday here in the US Th).

**Comment by lyschoening** [comment]
@boydgreenfield Looks great! Thank you for your work.

We can switch the endpoints to underscore notation. 

I've now committed a change that drops `App.potion`. This should enable support of multiple Blueprints. I had been considering for a while to remove it. It is only needed in rare cases, such as when a `ToOne()` field uses a string reference and is used within the application context but not part of a resource. In those cases it is virtually always possible to either make a specific reference to the resource class or provide a module path to the resource. 

There should be a test with a Blueprint that has a route prefix. There may be some further changes needed to properly accommodate those.

**Comment by TomNeyland** [comment]
@boydgreenfield I [opened a PR](https://github.com/refgenomics/potion/pull/1) against the branch you staged this PR from with an additional test case. The test covers behavior that I believe should be supported with blueprints, specifically the ability to register multiple api blueprints on a single flask app. 

@lyschoening Can you check out my comments on the above PR?

tldr: You cannot register multiple blueprints because potion will attempt to register the top level ```/schema/``` route for each blueprint, causing an assertion error.

**Comment by lyschoening** [comment]
@TomNeyland @boydgreenfield 

[From the documentation](http://flask.pocoo.org/docs/0.10/blueprints/#my-first-blueprint):

> Additionally it will prefix the endpoint of the function with the name of the blueprint which was given to the Blueprint constructor (in this case also simple_page).

The problem starts here: https://github.com/refgenomics/potion/commit/6afb392ee8818301f02a9b2da1c3f365109e3702#diff-5c8f369d1a696f426d314541b5de52d9R91
And continues here: https://github.com/refgenomics/potion/commit/6afb392ee8818301f02a9b2da1c3f365109e3702#diff-5c8f369d1a696f426d314541b5de52d9R101

`Flask.add_url_rule()` is called, but `Blueprint.add_url_rule()` should be called instead, as it would handle the prefixing: https://github.com/mitsuhiko/flask/blob/master/flask/blueprints.py#L75

**Comment by boydgreenfield** [comment]
@lyschoening I'm out for a few hours but can review your PR and look into this further. The way this is done in `Flask-Restful` may be a good model to follow, specifically: https://github.com/flask-restful/flask-restful/blob/master/flask_restful/__init__.py#L137-L186

**Comment by lyschoening** [comment]
@boydgreenfield Flask-RESTful is generally a good reference (the predecessor of Flask-Potion was built on top of RESTful) – it's worth noting, though, that in this case the RESTful code is only so complicated because they're patching Blueprint to build their url by passing a callable as `rule` to `add_url_rule()`. I am not sure if that complexity is required in Potion. It _might_ even be enough to simply pass the blueprint to `_init_app()`.

That being said, in Potion `Resource.route_prefix` has to be the complete rule prefix, so here you will likely have to prepend `self.blueprint.url_prefix` if set: https://github.com/biosustain/potion/blob/master/flask_potion/__init__.py#L184

**Comment by boydgreenfield** [comment]
Ok, I've rebased on master here and this now supports multiple Blueprints w/ prefixes, as far as I can tell (have added test cases here too). 

@lyschoening – One remaining design question here is how to best dispatch the `add_url_rule` call. Here I'm just checking for a `Api.blueprint` attribute and calling off `blueprint` or `app` in an if/else clause. Another pattern I've seen is to do `Api.app = blueprint` here too, but I'm not sure if that introduces further complexity / other edge cases. Lmk your preferences.

**Comment by lyschoening** [comment]
@boydgreenfield This looks fine other than the issue of `Resource.route_prefix`. `Resource.route_prefix` is used in the Api schema and in `ToOne`/`ToMany` fields to refer to a resource URI.

The callback for registering a Blueprint with an app takes a `url_prefix` argument. e.g.: `self.app.register_blueprint(api_bp, url_prefix='/blueprint-prefix')`. Any resource in the Api would then have the prefix `api_bp.url_prefix + api.prefix`. Blueprints will need to be treated differently than Apps because of this. 

This shouldn't be too difficult to implement. Just need to defer setting `Resource.route_prefix` until the blueprint is registered with an app and concatenating with the `url_prefix`. I will have a look tomorrow at adding that in, and then I think we can merge this.

**Comment by boydgreenfield** [comment]
@lyschoening Ah, I'm not using Blueprints in that way (with a url_prefix), but yes indeed that's a problem. cc2ef4d fixes this (and includes a test case)... but this is all starting to feel rather inelegant. 

Let me know if you prefer to merge this or tackle this in a cleaner way. Again, first time dealing with Flask app and Blueprint internal state in any detail, so I'll defer to you on preferred patterns here.

**Comment by lyschoening** [comment]
I've now merged this and refactored it a bit more. Included in 0.11.1. Thanks again for your work.

**Comment by boydgreenfield** [comment]
@lyschoening Great, thanks! And appreciate the extra cleanup / refactor too. :)
